### PR TITLE
Editor Tracking - Track context of block events (template parts, reusable blocks, etc.)

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
#### Changes proposed in this Pull Request

Adds a `entity_context` property to block events when fired from within an entity controlling container.  This will help track how/what people are editing in the site editor.  This also extends to the post editor for reusable blocks.

Currently, the types of contexts in this PR are: "template", "core/template-part", "core/post-content", "core/query", and "core/block" (the reusable block).  When a block event happens, we use the `rootClientId` for the event and bubble up through the block hierarchy looking for one of these entity controller blocks listed above.  The first is returned as the context, and if none is found the default context is determined to be at the top level and is marked as "template" for the site editor and left `undefined` for the post editor as this would be a repeat of the already existing `post_type` property.

This should work for the following events:
* Move Up / Move Down
* Removing blocks
* Inserting blocks / patterns
* replacing blocks
* <s>drag and drop</s> - while I set this up to also work for the current drag-and-drop event tracking and allowed for both `from` and `to` type contexts, I found that drag-and-drop events are no longer tracking as expected in its current implementation. (opened an issue here https://github.com/Automattic/wp-calypso/issues/54353)

For Template Parts, if a variation is active it is appended to the end of the context string (ex. `core/template-part/header`) and a separate `template_part_id` property is also added to the event.

#### Testing instructions

* Load the site editor - test various block events in various contexts and verify their `entity_context` is listed as expected above.  For the top level, verify the context is "template".
* Load the post editor - add a reusable block or template part.  Verify block events within these containers populate the `entity_context` property as expected.  For block events the top level, verify there is no `entity_context` property.

Related to https://github.com/Automattic/wp-calypso/issues/53410

### CODE REVIEW

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js [hunk 1]
```diff
@@ -204,12 +204,29 @@ function trackBlocksHandler( blocks, eventName, propertiesHandler = noop, parent
...
+
+	let context;
+	let from_context;
+	let to_context;
+	if ( toRootClientId === undefined || fromContext === toContext ) {
+		context = fromContext;
+	} else {
+		from_context = fromContext;
+		to_context = toContext;
+	}
```
Comment (Addison-Stavlo) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js]: Technically this extra complexity for from/to is only valid for tracking the `moveBlockToPosition` action.  Currently this is supposed to track reusable blocks but is not (see issue linked in PR description).  However, this event is being fired on rare occasions by another means (although im not sure where to find it?  block selection button keyDown events) so may be worth keeping anyways.

File: test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js [hunk 2]
```diff
@@ -1070,8 +1070,70 @@ describe( `[${ host }] Calypso Gutenberg Site Editor Tracking: (${ screenSize })
...
+				);
+				await driverHelper.setWhenSettable(
+					this.driver,
+					quickInserterSearchInputLocator,
+					'Heading'
+				);
+				await driverHelper.clickWhenClickable( this.driver, blockItemLocator );
 
-			await deleteCustomEntities( this.driver, 'wp_template_part' );
+				await editor.runInCanvas( async () => {} );
```
Comment (WunderBart) [inline test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js]: Is this necessary?

Comment (Addison-Stavlo) [reply test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js]: Oh no, i forgot to remove this.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js (hunk 1 above)
Comment (Addison-Stavlo) [reply apps/wpcom-block-editor/src/wpcom/features/tracking.js]: I looks like the block selection button is this which is found in select mode:

Screen Shot 2021-07-09 at 12 42 42 PM

Im not sure how to get its `moveBlocksToPosition` action to fire.

File: apps/wpcom-block-editor/src/wpcom/features/tracking.js [hunk 3]
```diff
@@ -208,13 +212,38 @@ function trackBlocksHandler( blocks, eventName, propertiesHandler = noop, parent
...
+		} );
+	} else {
+		const fromProps = {};
+		const toProps = {};
+		for ( const key of Object.keys( fromContext ) ) {
+			fromProps[ `from_${ key }` ] = fromContext[ key ];
+		}
+		for ( const key of Object.keys( toContext ) ) {
+			toProps[ `from_${ key }` ] = toContext[ key ];
+		}
```
Comment (Addison-Stavlo) [inline apps/wpcom-block-editor/src/wpcom/features/tracking.js:237]: still the same notes as previously listed https://github.com/Automattic/wp-calypso/pull/54352#discussion_r665544515

File: test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js [hunk 4]
```diff
@@ -1070,8 +1070,91 @@ describe( `[${ host }] Calypso Gutenberg Site Editor Tracking: (${ screenSize })
...
+				);
+				await driverHelper.clickWhenClickable(
+					this.driver,
+					By.css( '.block-editor-block-mover-button.is-down-button' )
+				);
+
+				const events = await getEventsStack( this.driver );
+				assert.strictEqual( events.length, 2 );
 
-			await deleteCustomEntities( this.driver, 'wp_template_part' );
```
Comment (Addison-Stavlo) [inline test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js:1131]: We can get rid of this from where it was.  Its not necessary to clean the template parts at that step in order for that test to pass again if ran.  Therefore, we can save this deletion until the normal cleanup stage to reduce its chance of conflicting with another parallel test run on this site.

**Review by david-szabo97 — APPROVED** [review]
E2E tests are passing and found no issues during testing 👍

**Review by mattwiebe — APPROVED** [review]
Tested in both post and site editors, works like it's supposed to. :shipit:

**Pull request merged**